File: indian-railways-geofence/static/js/distance.js
/**
 * Distance calculation utilities for the Indian Railways Geo-Fencing System
 * 
 * This file contains JavaScript implementations of the Haversine formula
 * and other geo-spatial calculations used in the system.
 */

// Calculate the great circle distance between two points using the Haversine formula
function haversineDistance(lat1, lon1, lat2, lon2) {
    // Convert latitude and longitude from degrees to radians
    lat1 = toRadians(lat1);
    lon1 = toRadians(lon1);
    lat2 = toRadians(lat2);
    lon2 = toRadians(lon2);
    
    // Haversine formula
    const dlon = lon2 - lon1;
    const dlat = lat2 - lat1;
    const a = Math.sin(dlat/2)**2 + Math.cos(lat1) * Math.cos(lat2) * Math.sin(dlon/2)**2;
    const c = 2 * Math.asin(Math.sqrt(a));
    
    // Earth's radius in kilometers
    const r = 6371;
    
    return c * r;
}

// Calculate the bearing between two points
function bearingBetweenPoints(lat1, lon1, lat2, lon2) {
    // Convert latitude and longitude from degrees to radians
    lat1 = toRadians(lat1);
    lon1 = toRadians(lon1);
    lat2 = toRadians(lat2);
    lon2 = toRadians(lon2);
    
    // Calculate bearing
    const y = Math.sin(lon2 - lon1) * Math.cos(lat2);
    const x = Math.cos(lat1) * Math.sin(lat2) - 
              Math.sin(lat1) * Math.cos(lat2) * Math.cos(lon2 - lon1);
    
    let bearing = Math.atan2(y, x);
    
    // Convert from radians to degrees
    bearing = toDegrees(bearing);
    
    // Normalize to 0-360
    bearing = (bearing + 360) % 360;
    
    return bearing;
}

// Calculate the destination point given a starting point, bearing, and distance
function destinationPoint(lat, lon, bearing, distance) {
    // Convert to radians
    lat = toRadians(lat);
    lon = toRadians(lon);
    bearing = toRadians(bearing);
    
    // Earth's radius in kilometers
    const R = 6371;
    
    // Angular distance
    const d = distance / R;
    
    // Calculate destination point
    const lat2 = Math.asin(Math.sin(lat) * Math.cos(d) + 
                         Math.cos(lat) * Math.sin(d) * Math.cos(bearing));
    
    const lon2 = lon + Math.atan2(Math.sin(bearing) * Math.sin(d) * Math.cos(lat),
                               Math.cos(d) - Math.sin(lat) * Math.sin(lat2));
    
    // Convert back to degrees
    return {
        lat: toDegrees(lat2),
        lon: toDegrees(lon2)
    };
}

// Check if a point is inside a geo-fence (circular)
function isPointInGeoFence(pointLat, pointLon, centerLat, centerLon, radiusKm) {
    const distance = haversineDistance(pointLat, pointLon, centerLat, centerLon);
    return distance <= radiusKm;
}

// Convert degrees to radians
function toRadians(degrees) {
    return degrees * Math.PI / 180;
}

// Convert radians to degrees
function toDegrees(radians) {
    return radians * 180 / Math.PI;
}

// Export functions for use in other scripts
window.haversineDistance = haversineDistance;
window.bearingBetweenPoints = bearingBetweenPoints;
window.destinationPoint = destinationPoint;
window.isPointInGeoFence = isPointInGeoFence;

File: indian-railways-geofence/app/utils/distance.py
import math

def haversine_distance(lat1, lon1, lat2, lon2):
    """
    Calculate the great circle distance between two points 
    on the earth (specified in decimal degrees)
    
    Returns distance in kilometers
    """
    # Convert latitude and longitude from degrees to radians
    lat1, lon1, lat2, lon2 = map(math.radians, [lat1, lon1, lat2, lon2])
    
    # Haversine formula
    dlon = lon2 - lon1
    dlat = lat2 - lat1
    a = math.sin(dlat/2)**2 + math.cos(lat1) * math.cos(lat2) * math.sin(dlon/2)**2
    c = 2 * math.asin(math.sqrt(a))
    
    # Earth's radius in kilometers
    r = 6371
    
    return c * r

def bearing_between_points(lat1, lon1, lat2, lon2):
    """
    Calculate the bearing between two points on Earth
    
    Returns bearing in degrees (0-360, where 0 is North)
    """
    # Convert latitude and longitude from degrees to radians
    lat1, lon1, lat2, lon2 = map(math.radians, [lat1, lon1, lat2, lon2])
    
    # Calculate bearing
    y = math.sin(lon2 - lon1) * math.cos(lat2)
    x = math.cos(lat1) * math.sin(lat2) - \
        math.sin(lat1) * math.cos(lat2) * math.cos(lon2 - lon1)
    
    bearing = math.atan2(y, x)
    
    # Convert from radians to degrees
    bearing = math.degrees(bearing)
    
    # Normalize to 0-360
    bearing = (bearing + 360) % 360
    
    return bearing

def destination_point(lat, lon, bearing, distance):
    """
    Calculate the destination point given a starting point, 
    bearing (in degrees), and distance (in kilometers)
    
    Returns (latitude, longitude) of destination point
    """
    # Convert to radians
    lat = math.radians(lat)
    lon = math.radians(lon)
    bearing = math.radians(bearing)
    
    # Earth's radius in kilometers
    R = 6371
    
    # Angular distance
    d = distance / R
    
    # Calculate destination point
    lat2 = math.asin(math.sin(lat) * math.cos(d) + 
                     math.cos(lat) * math.sin(d) * math.cos(bearing))
    
    lon2 = lon + math.atan2(math.sin(bearing) * math.sin(d) * math.cos(lat),
                           math.cos(d) - math.sin(lat) * math.sin(lat2))
    
    # Convert back to degrees
    lat2 = math.degrees(lat2)
    lon2 = math.degrees(lon2)
    
    return (lat2, lon2)